### PR TITLE
AI-1167: shared code support for SQL transformation tooling

### PR DESCRIPTION
**Pull request by Matovidlo**
## Description

**Linear**: [AI-1167](https://linear.app/keboola/issue/AI-1167/add-support-for-shared-codes-to-sql-transformation-tooling)

### Change Type

- [ ] Major (breaking changes, significant new features)
- [x] Minor (new features, enhancements, backward compatible)
- [ ] Patch (bug fixes, small improvements, no new features)

### Summary

Implements the RFC at `feature_spec/shared_code_support/RFC.md` (commit `e33f6ccb`). The MCP server can now discover, create, link and unlink Keboola **shared code** libraries — reusable SQL/Python/R snippets referenced from transformations via Mustache placeholders (`{{ rowId }}`).

**New surface (LLM-visible):**
- `get_shared_codes(transformation_component_ids=[...])` — read-only discovery tool; lists `keboola.shared-code` parent libraries and their rows.
- `add_config_row(..., row_id="<mustache-key>")` — optional explicit row ID forwarded as SAPI `rowId`. Required so the LLM can set Mustache placeholder keys on shared-code rows at creation time.
- `create_sql_transformation(..., shared_code_id="", shared_code_row_ids=[])` — links a new SQL transformation to a shared-code library at creation time.
- `update_sql_transformation(..., parameter_updates=[...])` — two new discriminated ops: `set_shared_code` (replaces linkage) and `remove_shared_code` (clears linkage). Both patch the configuration root, alongside `parameters`/`storage`.
- `create_config(..., shared_code_id="", shared_code_row_ids=[])` / `update_config(..., shared_code_id=..., shared_code_row_ids=...)` — same fields for Python/R/DuckDB transformations created via the generic config tools.

**Internal:**
- `TransformationConfiguration` model gains optional `shared_code_id` / `shared_code_row_ids` fields so existing linkage survives deserialize/re-serialize cycles.
- `configuration_row_create` storage-client method forwards `rowId` to SAPI (was missing).
- System prompt grows a **Shared Code** section between Transformations and Development Branches: covers discovery, when (not) to extract code, domain model, naming convention, the full CRUD pattern over existing generic tools, referencing mechanic, and validation rules (case sensitivity, bidirectional script/array consistency).

## Testing

- [x] Tested with Cursor AI desktop (`Streamable-HTTP` transports)

### Optional testing
- [ ] Tested with Cursor AI desktop (all transports)
- [ ] Tested with claude.ai web and `canary-orion` MCP (`Streamable-HTTP`)
- [ ] Tested with In Platform Agent on `canary-orion`
- [ ] Tested with RO chat on `canary-orion`

**Automated coverage:**
- **Unit tests** (`tests/tools/components/test_tools.py`): 16 new cases — rowId forwarding, root-field persistence at create, `TfSetSharedCode`/`TfRemoveSharedCode` dispatch, parameter-only update preservation, `create_config`/`update_config` set/clear/preserve flows, `get_shared_codes` happy path + filter + unknown-filter rejection.
- **Integration tests** (`integtests/tools/components/test_tools.py`): 7 new live-stack tests with a `shared_code_parent_factory` fixture for cleanup. Covers discovery (empty + filter), rowId forwarding to SAPI, create transformation with shared code, set/remove round-trip via `update_sql_transformation`, parameter-only update preservation (regression guard), `update_config` set/clear/preserve.
- **KaiBench Phase 9** ([keboola-rnd/KaiBench#23](https://github.com/keboola-rnd/KaiBench/pull/23)): MCP-09 (Snowflake), MCP-09-BQ (BigQuery), MCP-09-EDGE (edge cases) — end-to-end validation against the live MCP server.
- `tox` clean: flake8 OK, black OK, 991 unit tests pass, 37 integtests collect.

## Checklist

- [x] Self-review completed
- [x] Unit tests added/updated
- [x] Integration tests added/updated
- [x] Project version bumped (1.73.2 → 1.74.0, after merging main)
- [x] Documentation updated (system prompt, TOOLS.md regenerated)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[AI-1167]: https://keboola.atlassian.net/browse/AI-1167?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by Matovidlo** [comment]
@claude review

File: feature_spec/shared_code_support/RFC.md [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+Conventional config IDs (not enforced by the API, but used by default tooling):
+
+| Transformation type | Conventional shared-code config ID |
+|---|---|
+| `keboola.snowflake-transformation` | `shared-codes.snowflake-transformation` |
+| `keboola.google-bigquery-transformation` | `shared-codes.google-bigquery-transformation` |
+| `keboola.python-transformation-v2` | `shared-codes.python-transformation-v2` |
+| `keboola.duckdb-transformation` | `shared-codes.duckdb-transformation` |
+| `keboola.r-transformation-v2` | `shared-codes.r-transformation-v2` |
+
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply feature_spec/shared_code_support/RFC.md:53]: **Fixed** ✅ in commit e33f6ccb724ac4fcad1552889a9bb879b20b6824

Removed `keboola.duckdb-transformation` from all lists in the RFC (conventional config IDs table, `get_shared_codes` accepted values). No `DUCKDB_TRANSFORMATION_ID` constant exists in `utils.py` and no references appear under `src/`, so including it was premature. DuckDB support can be added in a follow-up once the constant is defined.

File: feature_spec/shared_code_support/RFC.md [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+sufficient for creating and editing shared code content. No new write tools are needed.
+The system prompt (section 1) must document the exact parameters so the LLM can use them
+correctly:
+
+| Operation | Tool | Key parameters |
+|---|---|---|
+| Create parent config | `create_config` | `component_id="keboola.shared-code"`, `parameters={"componentId":"<tf-component-id>"}` |
+| Add snippet row | `add_config_row` | `component_id="keboola.shared-code"`, `row_id="<mustache-key>"`, `parameters={"code_content":["<code>"]}` |
+| Update snippet | `update_config_row` | `parameter_updates=[{"op":"set","path":"code_content","value":["<new code>"]}]` |
+| Delete snippet | `delete_config_row` (if it exists) or `update_config_row` to disable | — |
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply feature_spec/shared_code_support/RFC.md]: **Fixed** ✅ in commit e33f6ccb724ac4fcad1552889a9bb879b20b6824

Replaced the `delete_config_row` reference with `update_config_row` + `is_disabled=True`. No `delete_config_row` tool exists in the codebase — only `add_config_row` and `update_config_row` are exported. The CRUD table now reflects what is actually available.

File: feature_spec/shared_code_support/RFC.md [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+
+```
+create_sql_transformation(
+  name="My Transformation",
+  sql_code_blocks=[
+    Block(name="Main", codes=[Code(name="Reused logic", script="{{ dumpfiles }}")])
+  ],
+  shared_code_id="shared-codes.snowflake-transformation",
+  shared_code_row_ids=["dumpfiles"]
+)
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply feature_spec/shared_code_support/RFC.md:442]: **Fixed** ✅ in commit e33f6ccb724ac4fcad1552889a9bb879b20b6824

Fixed the `create_sql_transformation` example: replaced `Block(name=..., codes=[Code(...)])` with a flat list `[Code(name=..., script=...)]`, matching the actual `Sequence[SimplifiedTfBlocks.Block.Code]` parameter type. The tool wraps the codes into a single block internally.

File: feature_spec/shared_code_support/RFC.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Matovidlo) [reply feature_spec/shared_code_support/RFC.md:53]: This may be incorrect if we want to supporty duck DB but no clue whether we want new tools or embed it to existing transfromation tools

[the remark this answers is not in the document]
Comment (Matovidlo) [reply feature_spec/shared_code_support/RFC.md:53]: **Fixed** ✅ in commit fab2b6086f22212d8c9a27afec3492fe4f993fc0

Removed DuckDB from the RFC's supported-transformation prose (sections 4c/4d/5). No `keboola.duckdb-transformation` constant exists in the codebase, so DuckDB is not actually supported; the DuckDB row was already gone from the config-IDs table. The historical changelog line for aa8fc636 is left as a record of that commit.

File: src/keboola_mcp_server/tools/components/tools.py [hunk 4]
```diff
@@ -1085,7 +1339,13 @@ async def create_config(
         initial_message='The "parameters" field is not valid.',
     )
 
-    configuration_payload = {'storage': storage_cfg, 'parameters': parameters}
+    # `keboola.shared-code` parent libraries use a flat configuration body — the platform's UI
+    # and runtime expansion read `componentId` at the configuration root, not under `parameters`.
+    # For every other component the generic wrapper applies.
+    if component_id == SHARED_CODE_COMPONENT_ID:
+        configuration_payload: dict[str, Any] = dict(parameters or {})
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py:1463]: **Fixed** ✅ in commit fab2b6086f22212d8c9a27afec3492fe4f993fc0

create_config now rejects a `keboola.shared-code` parent when `configuration_id` is empty, raising a ValueError before calling configuration_create — an auto-assigned UUID would create a library the UI/runtime expansion cannot resolve. Covered by new test test_create_config_shared_code_parent_requires_conventional_id.

File: src/keboola_mcp_server/tools/components/tools.py [hunk 5]
```diff
@@ -1230,7 +1522,12 @@ async def add_config_row(
         configuration_id=configuration_id,
     )
 
-    configuration_payload = {'storage': storage_cfg, 'parameters': parameters}
+    # `keboola.shared-code` rows use a flat configuration body — the platform reads
+    # `code_content` at the row configuration root, not under `parameters`.
+    if component_id == SHARED_CODE_COMPONENT_ID:
+        configuration_payload: dict[str, Any] = dict(parameters or {})
+    else:
+        configuration_payload = {'storage': storage_cfg, 'parameters': parameters}
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py:1654]: **Fixed** ✅ in commit fab2b6086f22212d8c9a27afec3492fe4f993fc0

update_config_row_internal now applies parameter_updates against the flat row root for `keboola.shared-code` (mirroring add_config_row), so editing `code_content` lands at the root where get_shared_codes reads it — no longer silently written to parameters.code_content. Covered by new test test_update_config_row_shared_code_applies_updates_to_flat_root.

File: src/keboola_mcp_server/tools/components/tools.py [hunk 6]
```diff
@@ -1040,6 +1248,42 @@ async def create_config(
...
     ] = None,
+    shared_code_id: Annotated[
+        str,
+        Field(
+            description=(
+                'Optional. The configuration ID of the parent `keboola.shared-code` library this configuration '
+                'references at the root level. Useful when creating Python (`keboola.python-transformation-v2`), '
+                'R (`keboola.r-transformation-v2`) or DuckDB transformation configurations that need to reuse '
+                'shared snippets. Must be paired with `shared_code_row_ids` and matching `{{ rowId }}` placeholders '
+                "in the component's script. Leave empty when not using shared code."
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py:1339]: **Fixed** ✅ in commit fab2b6086f22212d8c9a27afec3492fe4f993fc0

Removed DuckDB from the create_config `shared_code_id` description and docstring (and regenerated TOOLS.md). The allowlist SHARED_CODE_TRANSFORMATION_IDS is unchanged (Snowflake/BigQuery/Python/R); DuckDB is no longer advertised since no DuckDB component ID is defined.

File: integtests/tools/components/test_tools.py [hunk 7]
```diff
@@ -1050,3 +1060,383 @@ async def test_get_config_examples_with_invalid_component(mcp_context: Context):
...
+        description='Reusable snippet for integration tests',
+        component_id=SHARED_CODE_COMPONENT_ID,
+        configuration_id=parent.configuration_id,
+        parameters={'code_content': ['SELECT 1 AS integ_marker']},
+        row_id=row_id,
+    )
+
+    detail = await client.storage_client.configuration_detail(
+        component_id=SHARED_CODE_COMPONENT_ID,
+        configuration_id=parent.configuration_id,
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply integtests/tools/components/test_tools.py:1159]: **Fixed** ✅ in commit fab2b6086f22212d8c9a27afec3492fe4f993fc0

Added include=['rows'] to the configuration_detail call in the shared-code integration test (integtests/tools/components/test_tools.py) so SAPI returns row data before the test asserts on detail['rows'].

File: src/keboola_mcp_server/tools/components/utils.py [hunk 8]
```diff
@@ -65,6 +65,17 @@
...
+
+# Transformation component IDs that may have a shared-code library associated with them.
+SHARED_CODE_TRANSFORMATION_IDS: frozenset[str] = frozenset(
+    {
+        SNOWFLAKE_TRANSFORMATION_ID,
+        BIGQUERY_TRANSFORMATION_ID,
+        PYTHON_TRANSFORMATION_ID,
+        R_TRANSFORMATION_ID,
+    }
+)
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/utils.py:80]: **Fixed** ✅ in commit fab2b6086f22212d8c9a27afec3492fe4f993fc0

Removed DuckDB from the LLM-visible docs rather than expanding the allowlist. SHARED_CODE_TRANSFORMATION_IDS intentionally stays Snowflake/BigQuery/Python/R; DuckDB is no longer claimed anywhere in tool docs, TOOLS.md or the RFC prose, so there is no advertised-but-unsupported behaviour.

File: integtests/tools/components/test_tools.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Matovidlo) [reply integtests/tools/components/test_tools.py:1159]: **Fixed** ✅ in commit fab2b6086f22212d8c9a27afec3492fe4f993fc0

Same fix as the other line-1154 thread: the shared-code parent detail fetch now passes include=['rows'], so rowId forwarding is verified against actual row data instead of a possibly-empty list.

File: src/keboola_mcp_server/tools/components/tools.py [hunk 9]
```diff
@@ -373,6 +390,101 @@ async def fetch_component_with_links(component_id: str) -> Component:
...
+            component_id=SHARED_CODE_COMPONENT_ID,
+            configuration_id=str(parent.get('id')),
+            include=['rows'],
+        )
+        rows: list[SharedCodeRow] = []
+        for raw_row in cast(list[dict[str, Any]], detail.get('rows') or []):
+            row_config = cast(dict[str, Any], raw_row.get('configuration') or {})
+            row_parameters = cast(dict[str, Any], row_config.get('parameters') or {})
+            code_content = row_config.get('code_content')
+            if code_content is None:
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py:482]: **Fixed** ✅ in commit fab2b6086f22212d8c9a27afec3492fe4f993fc0

get_shared_codes now skips rows with isDisabled truthy, so disabled snippets (the documented update_config_row(is_disabled=True) soft-delete path) are no longer surfaced as reusable code. Covered by new test test_get_shared_codes_skips_disabled_rows.

File: src/keboola_mcp_server/tools/components/utils.py [hunk 10]
```diff
@@ -370,6 +381,219 @@ async def create_transformation_configuration(
...
+    for block in blocks or ():
+        codes = block.get('codes', []) if isinstance(block, dict) else getattr(block, 'codes', [])
+        for code in codes or ():
+            script = code.get('script') if isinstance(code, dict) else getattr(code, 'script', None)
+            items = script if isinstance(script, list) else [script]
+            for item in items:
+                if not isinstance(item, str):
+                    continue
+                referenced.update(SHARED_CODE_PLACEHOLDER_RE.findall(item))
+    return referenced
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/utils.py:414]: **Fixed** ✅ in commit a8eab88f91f20a1d1d5b082035ae2c821d9db5f4

Good catch. extract_shared_code_row_ids_from_blocks now matches a placeholder only when it is the sole content of a script element (PURE_SHARED_CODE_PLACEHOLDER_RE.match), the only form the platform substitutes for shared code. Inline occurrences like `token={{ api_token }}` — Keboola configuration variables sharing the same Mustache syntax — are no longer treated as shared-code references, so validate_shared_code_linkage no longer blocks transformations that legitimately use variables. Added regression test test_create_sql_transformation_allows_inline_config_variable_without_linkage; the existing reject tests still pass because they use whole-element placeholders.

File: integtests/tools/components/test_tools.py [hunk 11]
```diff
@@ -1050,3 +1060,384 @@ async def test_get_config_examples_with_invalid_component(mcp_context: Context):
...
+    created: list[tuple[str, str]] = []  # (component_id, configuration_id)
+
+    async def make(target_component_id: str, name: str = '') -> ConfigToolOutput:
+        config = await create_config(
+            ctx=mcp_context,
+            name=name or f'Shared Codes for {target_component_id}',
+            description='Created by integtest — safe to delete',
+            component_id=SHARED_CODE_COMPONENT_ID,
+            parameters={'componentId': target_component_id},
+        )
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply integtests/tools/components/test_tools.py:1091]: **Fixed** ✅ in commit a8eab88f91f20a1d1d5b082035ae2c821d9db5f4

Fixed — the shared_code_parent_factory fixture now passes the conventional configuration_id (shared-codes.<transformation-component-id>, derived from the target component id), matching the new create_config requirement so the integration tests keep working.

File: src/keboola_mcp_server/tools/components/tools.py [hunk 12]
```diff
@@ -480,6 +623,26 @@ async def create_sql_transformation(
     transformation_configuration_payload = await create_transformation_configuration(
         codes=sql_code_blocks, transformation_name=name, output_tables=created_table_names, sql_dialect=sql_dialect
     )
+    if shared_code_id:
+        transformation_configuration_payload.shared_code_id = shared_code_id
+        transformation_configuration_payload.shared_code_row_ids = list(shared_code_row_ids)
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py:639]: **Fixed** ✅ in commit f79f502924384a0293e5b041a9ea7fc37c443561

Added validate_shared_code_params: create_sql_transformation (and create_config) now raise when shared_code_row_ids is provided without a shared_code_id, instead of silently dropping the row IDs. update_config intentionally keeps its documented three-state semantics (empty string clears the linkage, None leaves it untouched), so the guard is applied only to the create-time tools. Covered by new test test_create_sql_transformation_rejects_row_ids_without_shared_code_id.

File: src/keboola_mcp_server/tools/components/utils.py [hunk 13]
```diff
@@ -370,6 +381,224 @@ async def create_transformation_configuration(
...
+
+    if not blocks:
+        if not row_ids:
+            return
+        blocks.append({'name': 'Blocks', 'codes': []})
+
+    target_block = blocks[0]
+    codes = target_block.setdefault('codes', [])
+    user_codes = [c for c in codes if not is_shared_code_marker(c.get('name', ''), c.get('script'))]
+    marker_codes: list[dict[str, Any]] = []
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/utils.py:576]: **Fixed** ✅ in commit f79f502924384a0293e5b041a9ea7fc37c443561

sync_shared_code_markers_in_dict now strips existing `Shared Code (...)` marker blocks from EVERY block before re-inserting the canonical set into the first block, so a marker moved into another block during manual reordering can no longer leave a stale copy that executes the snippet twice. Covered by new test test_sync_shared_code_markers_strips_markers_from_all_blocks.

File: src/keboola_mcp_server/tools/components/tools.py [hunk 14]
```diff
@@ -1882,17 +2254,30 @@ async def update_config_row_internal(
...
+            # `parameters`, and `get_shared_codes` reads the root first. Apply updates against the
+            # root (mirroring `add_config_row`) so an edit to e.g. `code_content` is actually visible.
+            updated_root = update_params(configuration_payload, parameter_updates)
+            configuration_payload = validate_row_parameters_configuration(
+                component=component,
+                parameters=updated_root,
+                initial_message='Applying the "parameter_updates" resulted in an invalid row configuration.',
+                configuration_id=configuration_id,
+                configuration_row_id=configuration_row_id,
+            )
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py:2311]: **Fixed** ✅ in commit bc035cdd2a132e11876fe7cbc88dff912b930820

Good catch. In the shared-code branch of update_config_row_internal I now hold aside the structural `storage`/`processors` siblings before validation, validate only the flat parameter body (code_content etc.) against the row schema, and re-attach the siblings afterwards — so they are neither dropped by validate_row_parameters_configuration (which returns only the parameter body) nor wrongly validated against the parameter schema. Extended test_update_config_row_shared_code_applies_updates_to_flat_root to assert a `storage` sibling survives the update.

File: src/keboola_mcp_server/resources/prompts/project_system_prompt.md [hunk 15]
```diff
@@ -66,6 +66,44 @@ See the [Development Branches](#development-branches) section for more details.
...
+**Hard rules** — the tool will reject the call otherwise:
+
+- **Row content must be a complete executable statement** (e.g. `CREATE OR REPLACE TABLE …`, `from datetime import …`).
+  The runtime substitutes a `{{ rowId }}` script array element with the row's `code_content` array and runs the
+  result as its own query. Fragments (column-list, `WHERE` clause, sub-expression) will fail at run time.
+- **A `{{ rowId }}` placeholder requires linkage** — every referenced row must appear in `shared_code_row_ids` and
+  `shared_code_id` must be set. Without the linkage the platform cannot resolve the placeholder. The tool emits
+  one marker block per linked row (or skips it when your own code already has `["{{ rowId }}"]` as a standalone
+  script element — no duplicate); **do NOT embed `{{ rowId }}` inline inside another SQL string** — inline
+  placeholders are not substituted (only `["{{ rowId }}"]` as its own array element is).
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/resources/prompts/project_system_prompt.md]: **Fixed** ✅ in commit a22e49df2b74ed74553f5f4f4c49869a729f3cff

Reworded the section. It's now split into 'Tool-enforced rules' (pure-placeholder linkage, shared_code_row_ids-requires-shared_code_id, case sensitivity) and 'Runtime requirements' (snippet completeness and pure-element-only substitution). Snippet completeness and inline {{ rowId }} handling are now described as runtime behaviour, not tool-level rejection, matching the implementation.

File: src/keboola_mcp_server/resources/prompts/project_system_prompt.md [hunk 16]
```diff
@@ -66,6 +66,44 @@ See the [Development Branches](#development-branches) section for more details.
...
+- **Row IDs are case-sensitive.**
+- **Marker ordering** — the auto-emitted marker is appended *after* the user-authored code in
+  the first parameters block. If your user code depends on the shared snippet's side-effect
+  (e.g. shared code does `SET (region) = ('EU')` and your code reads `$region`, or shared code
+  creates a temp table/view your code selects from), the marker must run first. Reorder via
+  `update_sql_transformation` with `remove_code` + `add_code(position="start")`, or author the
+  marker yourself as a pure `["{{ rowId }}"]` script block at the start (the tool de-duplicates).
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/resources/prompts/project_system_prompt.md]: **Fixed** ✅ in commit a22e49df2b74ed74553f5f4f4c49869a729f3cff

Fixed the marker-ordering guidance. It now states that the sync strips ALL recognised marker blocks and re-appends the canonical set to the first block, so reordering an auto-emitted `Shared Code (...)` marker via remove_code/add_code will not persist. Instead it instructs authoring your own non-marker pure `["{{ rowId }}"]` code block at the desired position — the de-dup logic (user_substitution_eligible_row_ids) recognises it and skips emitting a duplicate marker for that row.

File: src/keboola_mcp_server/tools/components/tools.py [hunk 17]
```diff
@@ -1003,26 +1168,73 @@ async def update_sql_transformation_internal(
...
+                block_updates.append(update)
+
+        for update in root_updates:
+            if isinstance(update, TfSetSharedCode):
+                updated_configuration['shared_code_id'] = update.shared_code_id
+                updated_configuration['shared_code_row_ids'] = list(update.shared_code_row_ids)
+                shared_code_messages.append(
+                    f'Linked shared code {update.shared_code_id!r} '
+                    f'(rows: {", ".join(update.shared_code_row_ids) or "<none>"}).'
+                )
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py:1207]: **Fixed** ✅ in commit dc991fbd489383fba7493d71924cea2064656297

The set_shared_code branch in update_sql_transformation_internal now raises when shared_code_id is empty, instructing callers to use remove_shared_code to clear linkage instead of writing an inconsistent empty-id root. Covered by new test test_update_sql_transformation_set_shared_code_rejects_empty_id.

File: pyproject.toml [hunk 18]
```diff
@@ -4,7 +4,7 @@ build-backend = "hatchling.build"
 
 [project]
 name = "keboola-mcp-server"
-version = "1.73.2"
+version = "1.74.0"
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply pyproject.toml:7]: **Fixed** ✅ in commit dc991fbd489383fba7493d71924cea2064656297

Updated the PR description checklist to reflect the real bump (1.73.2 → 1.74.0, after merging main into this branch). The 1.60.0 → 1.61.0 line was from before the merge, when main was still on the older version.

File: src/keboola_mcp_server/tools/components/tools.py [hunk 19]
```diff
@@ -1158,7 +1422,23 @@ async def create_config(
...
+        # an orphaned library that never resolves. Require the caller to pass it explicitly.
+        if not configuration_id:
+            raise ValueError(
+                'Creating a `keboola.shared-code` library requires an explicit `configuration_id` using the '
+                'conventional `shared-codes.<transformation-component-id>` value (e.g. '
+                '`shared-codes.snowflake-transformation`). Auto-assigned IDs are not recognised by the runtime '
+                'expansion or the UI.'
+            )
+        configuration_payload: dict[str, Any] = dict(parameters or {})
+    else:
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py:1464]: **Fixed** ✅ in commit f510db8f834249ad1e50b9e68b990a2c1f066ad2

create_config now derives the expected conventional ID (shared-codes.<suffix>) from parameters.componentId and rejects any configuration_id that doesn't match it exactly, before calling SAPI. A typo/non-conventional ID is caught with a message naming the correct ID. Covered by new test test_create_config_shared_code_parent_rejects_nonconventional_id (the existing conventional-ID happy-path test still passes).

File: src/keboola_mcp_server/tools/components/tools.py [hunk 20]
```diff
@@ -1335,9 +1653,21 @@ async def add_config_row(
...
+            'Use the assigned id when addressing this row.',
+            row_id,
+            assigned_row_id,
+            component_id,
+            configuration_id,
+        )
+
     LOG.info(
         f'Created new configuration for component "{component_id}" with configuration id ' f'"{configuration_id}".'
     )
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py:1698]: **Fixed** ✅ in commit f510db8f834249ad1e50b9e68b990a2c1f066ad2

Fixed the log line: add_config_row now logs 'Created new configuration row "<row-id>" for component ... in configuration ...', including the assigned row id, instead of implying a whole configuration was created.

File: src/keboola_mcp_server/tools/components/utils.py [hunk 21]
```diff
@@ -370,6 +381,244 @@ async def create_transformation_configuration(
...
+    if not isinstance(code_name, str) or not code_name.startswith('Shared Code ('):
+        return False
+    script_text: str
+    if isinstance(code_script, list):
+        if len(code_script) != 1:
+            return False
+        script_text = str(code_script[0])
+    else:
+        script_text = str(code_script or '')
+    return script_text.strip().startswith('{{') and script_text.strip().endswith('}}')
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/utils.py]: **Fixed** ✅ in commit d110b3b63e0a34ed21c092da512353efa9d3b7ec

is_shared_code_marker now requires the single script element to fully match PURE_SHARED_CODE_PLACEHOLDER_RE instead of the loose startswith('{{')/endswith('}}') check, so a multi-placeholder element like '{{ a }}\n{{ b }}' (or a placeholder wrapped in SQL) is no longer misclassified as a marker and stripped during sync. Covered by new parametrized test test_is_shared_code_marker.

File: src/keboola_mcp_server/tools/components/tools.py [hunk 22]
```diff
@@ -378,6 +396,105 @@ async def fetch_component_with_links(component_id: str) -> Component:
...
+        # `code_content` live at the configuration root, not under `parameters`. Historical
+        # configs created via the generic create_config wrapper may still have them nested,
+        # so read root first and fall back to `parameters.<field>` for compatibility.
+        parent_configuration = cast(dict[str, Any], parent.get('configuration') or {})
+        parent_parameters = cast(dict[str, Any], parent_configuration.get('parameters') or {})
+        transformation_component_id = cast(
+            str, parent_configuration.get('componentId') or parent_parameters.get('componentId') or ''
+        )
+        if filter_set and transformation_component_id not in filter_set:
+            return None
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py:466]: **Fixed** ✅ in commit 99e82468491674344d4f5b1bc669c3d5d01e372b

build_shared_code_config now logs a warning and returns None (skips the parent) when transformation_component_id is empty — i.e. the shared-code parent has no componentId at the root or under parameters. get_shared_codes therefore never emits a SharedCodeConfig with an empty transformation_component_id. Covered by new test test_get_shared_codes_skips_config_without_component_id.

File: src/keboola_mcp_server/tools/components/tools.py [hunk 23]
```diff
@@ -1158,7 +1431,34 @@ async def create_config(
...
+    # For every other component the generic wrapper applies.
+    if component_id == SHARED_CODE_COMPONENT_ID:
+        # The UI and runtime expansion only recognise shared-code libraries by the conventional
+        # `shared-codes.<transformation-component-id>` ID — a SAPI-auto-assigned UUID (or a typo)
+        # would create an orphaned library that never resolves. Require the caller to pass the
+        # exact conventional ID derived from the target `componentId`.
+        target_component_id = str((parameters or {}).get('componentId') or '')
+        expected_configuration_id = (
+            f'shared-codes.{target_component_id.split(".", 1)[-1]}' if target_component_id else ''
+        )
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py]: **Fixed** ✅ in commit 19d78ac8d58179de804c2dd014202357dbf7e356

create_config now rejects a keboola.shared-code parent when parameters.componentId is empty, before calling SAPI — such a config has no backend association and would later be skipped by get_shared_codes. (The conventional-ID derivation and match check now always run, since componentId is guaranteed present.) Covered by new test test_create_config_shared_code_parent_requires_component_id.

File: src/keboola_mcp_server/tools/components/tools.py [hunk 24]
```diff
@@ -1882,17 +2281,36 @@ async def update_config_row_internal(
...
+            siblings = {key: updated_root[key] for key in ('storage', 'processors') if key in updated_root}
+            flat_body = {key: value for key, value in updated_root.items() if key not in siblings}
+            validated_flat = validate_row_parameters_configuration(
+                component=component,
+                parameters=flat_body,
+                initial_message='Applying the "parameter_updates" resulted in an invalid row configuration.',
+                configuration_id=configuration_id,
+                configuration_row_id=configuration_row_id,
+            )
+            configuration_payload = {**validated_flat, **siblings}
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/components/tools.py:2312]: **Fixed** ✅ in commit 19d78ac8d58179de804c2dd014202357dbf7e356

The shared-code row update path now flattens a legacy nested `parameters` wrapper into the flat root before validation: it pops flat_body['parameters'] and merges it under the root (root wins). Because validate_row_parameters_configuration unwraps a `parameters` key and returns only its contents, without this a flat-root code_content update would be silently dropped for legacy wrapper-created rows. Covered by new test test_update_config_row_shared_code_flattens_legacy_parameters_wrapper.
